File: libraries/java/src/main/java/com/bhf/aeroncache/client/EmbeddedCounterCache.java
package com.bhf.aeroncache.client;

import com.bhf.aeroncache.models.*;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

/**
 * A counter cache view that mirrors remote updates into a local map. It works over either transport via
 * {@link CacheTransport}: the HTTP+WS {@link AeronCacheClient} or the Aeron
 * {@link com.bhf.aeroncache.client.gateway.AeronGatewayClient}.
 * <p>
 * Two subscription styles are offered: the transport-neutral {@link #subscribe(Consumer)} (works over
 * both transports), and the WebSocket-listener {@link #subscribe(CounterCacheSubscriber)} (HTTP only,
 * retained for backwards compatibility).
 */
public class EmbeddedCounterCache {
    private final CacheTransport transport;
    /** Non-null only when backed by the HTTP client; enables the legacy WebSocket-listener subscribe API. */
    private final AeronCacheClient httpClient;
    private final String cacheId;
    private final Map<String, Long> localCache = new ConcurrentHashMap<>();

    public EmbeddedCounterCache(AeronCacheClient client, String cacheId) {
        this.transport = client;
        this.httpClient = client;
        this.cacheId = cacheId;
    }

    public EmbeddedCounterCache(CacheTransport transport, String cacheId) {
        this.transport = transport;
        this.httpClient = transport instanceof AeronCacheClient ac ? ac : null;
        this.cacheId = cacheId;
    }

    public Long getLocal(String key) {
        return localCache.get(key);
    }

    public Map<String, Long> getLocalCache() {
        return java.util.Collections.unmodifiableMap(localCache);
    }

    public PutItemResponse put(String key, long value) throws Exception {
        return transport.putCounter(cacheId, key, value);
    }

    public PutItemResponse putTimed(String key, long value, long ttl) throws Exception {
        return transport.putTimedCounter(cacheId, key, value, ttl);
    }

    public CounterResponse get(String key) throws Exception {
        return transport.getCounter(cacheId, key);
    }

    public CounterResponse increment(String key, long amount) throws Exception {
        return transport.incrementCounter(cacheId, key, amount);
    }

    public CounterResponse decrement(String key, long amount) throws Exception {
        return transport.decrementCounter(cacheId, key, amount);
    }

    public CounterResponse set(String key, long value) throws Exception {
        return transport.setCounter(cacheId, key, value);
    }

    public DeleteItemResponse remove(String key) throws Exception {
        return transport.deleteCounter(cacheId, key);
    }

    public DeleteCacheResponse clear() throws Exception {
        return transport.deleteCounterCache(cacheId);
    }

    public CompletableFuture<PutItemResponse> putAsync(String key, long value) {
        return transport.putCounterAsync(cacheId, key, value);
    }

    public CompletableFuture<PutItemResponse> putTimedAsync(String key, long value, long ttl) {
        return transport.putTimedCounterAsync(cacheId, key, value, ttl);
    }

    public CompletableFuture<CounterResponse> getAsync(String key) {
        return transport.getCounterAsync(cacheId, key);
    }

    public CompletableFuture<CounterResponse> incrementAsync(String key, long amount) {
        return transport.incrementCounterAsync(cacheId, key, amount);
    }

    public CompletableFuture<CounterResponse> decrementAsync(String key, long amount) {
        return transport.decrementCounterAsync(cacheId, key, amount);
    }

    public CompletableFuture<CounterResponse> setAsync(String key, long value) {
        return transport.setCounterAsync(cacheId, key, value);
    }

    public CompletableFuture<DeleteItemResponse> removeAsync(String key) {
        return transport.deleteCounterAsync(cacheId, key);
    }

    public CompletableFuture<DeleteCacheResponse> clearAsync() {
        return transport.deleteCounterCacheAsync(cacheId);
    }

    // --- Transport-neutral subscription (works over HTTP+WS and Aeron) ---

    /**
     * Subscribe to counter updates, mirroring them into the local map and forwarding each event to
     * {@code listener}.
     *
     * @return a handle that unsubscribes when {@link AutoCloseable#close() closed}.
     */
    public AutoCloseable subscribe(Consumer<CounterUpdateEvent> listener) {
        return subscribe(listener, false);
    }

    public AutoCloseable subscribe(Consumer<CounterUpdateEvent> listener, boolean hydrate) {
        return transport.subscribeCounterUpdates(cacheId, hydrate, event -> {
            updateLocalCache(event);
            if (listener != null) {
                listener.accept(event);
            }
        });
    }

    // --- WebSocket-listener subscription (HTTP transport only, retained for compatibility) ---

    public ReconnectingWebSocket subscribe(CounterCacheSubscriber subscriber) {
        return subscribe(subscriber, false);
    }

    public ReconnectingWebSocket subscribe(CounterCacheSubscriber subscriber, boolean hydrate) {
        if (httpClient == null) {
            throw new UnsupportedOperationException(
                    "The WebSocket-listener subscribe API requires the HTTP transport; "
                            + "use subscribe(Consumer<CounterUpdateEvent>) with the Aeron transport.");
        }
        subscriber.setInternalUpdater(this::updateLocalCache);
        return httpClient.subscribeCounter(cacheId, hydrate, subscriber);
    }

    private void updateLocalCache(CounterUpdateEvent event) {
        String eventType = event.getEventType();
        String key = event.getItemKey();
        Long value = event.getItemValue();

        if ("ADD_ITEM".equals(eventType)) {
            if (key != null && value != null) {
                localCache.put(key, value);
            }
        } else if ("REMOVE_ITEM".equals(eventType)) {
            if (key != null) {
                localCache.remove(key);
            }
        } else if ("DELETE_CACHE".equals(eventType) || "CLEAR_CACHE".equals(eventType)) {
            localCache.clear();
        }
    }
}

File: libraries/python/aeron_cache/embedded_counter_cache.py

from .models import (
    PutItemResponse,
    DeleteItemResponse,
    DeleteCacheResponse,
    CounterResponse,
    CounterUpdateEvent
)

class EmbeddedCounterCache:
    def __init__(self, client, cache_id):
        self.client = client
        self.cache_id = cache_id
        self.local_cache = {}

    def get_local(self, key):
        return self.local_cache.get(key)

    def put(self, key, value) -> PutItemResponse:
        return self.client.put_counter(self.cache_id, key, value)

    def get(self, key) -> CounterResponse:
        return self.client.get_counter(self.cache_id, key)

    def increment(self, key, amount) -> CounterResponse:
        return self.client.increment_counter(self.cache_id, key, amount)

    def decrement(self, key, amount) -> CounterResponse:
        return self.client.decrement_counter(self.cache_id, key, amount)

    def set(self, key, value) -> CounterResponse:
        return self.client.set_counter(self.cache_id, key, value)

    def remove(self, key) -> DeleteItemResponse:
        return self.client.delete_counter(self.cache_id, key)

    def clear(self) -> DeleteCacheResponse:
        return self.client.delete_counter_cache(self.cache_id)

    async def put_async(self, key, value) -> PutItemResponse:
        return await self.client.put_counter_async(self.cache_id, key, value)

    async def get_async(self, key) -> CounterResponse:
        return await self.client.get_counter_async(self.cache_id, key)

    async def increment_async(self, key, amount) -> CounterResponse:
        return await self.client.increment_counter_async(self.cache_id, key, amount)

    async def decrement_async(self, key, amount) -> CounterResponse:
        return await self.client.decrement_counter_async(self.cache_id, key, amount)

    async def set_async(self, key, value) -> CounterResponse:
        return await self.client.set_counter_async(self.cache_id, key, value)

    async def remove_async(self, key) -> DeleteItemResponse:
        return await self.client.delete_counter_async(self.cache_id, key)

    async def clear_async(self) -> DeleteCacheResponse:
        return await self.client.delete_counter_cache_async(self.cache_id)

    async def subscribe(self, callback):
        async def wrapped_callback(event: CounterUpdateEvent):
            self._update_local_cache(event)
            await callback(event)
        return await self.client.subscribe_counter(self.cache_id, wrapped_callback)

    def _update_local_cache(self, event: CounterUpdateEvent):
        event_type = event.eventType
        if event_type == 'ADD_ITEM':
            if event.itemKey is not None and event.itemValue is not None:
                self.local_cache[event.itemKey] = event.itemValue
        elif event_type == 'REMOVE_ITEM':
            if event.itemKey is not None:
                self.local_cache.pop(event.itemKey, None)
        elif event_type in ['CLEAR_CACHE', 'DELETE_CACHE']:
            self.local_cache.clear()
